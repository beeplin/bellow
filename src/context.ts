import { AsyncLocalStorage } from 'async_hooks'
import type { Fn } from './types'

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export class ContextStore<C extends Record<string, any>> {
  private als = new AsyncLocalStorage<{ context: C }>()

  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  private fakeContext = {} as C

  set(value: C): void {
    const store = this.als.getStore()
    if (store) store.context = value
    else this.fakeContext = value
  }

  get(): C {
    return this.als.getStore()?.context ?? this.fakeContext
  }

  runWith(value: C, fn: Fn): void {
    this.als.run({ context: value }, fn)
  }
}
