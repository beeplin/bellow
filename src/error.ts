import type { Fn, TeleResponse } from './types'

export class TeleError<T extends Fn> extends Error {
  code: number

  data: unknown

  constructor(error: Required<TeleResponse<T>>['error']) {
    super()
    this.name = 'TeleError'
    this.code = error.code
    this.message = error.message
    this.data = error.data
  }
}
