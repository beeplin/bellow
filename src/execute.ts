import type {
  Fn,
  TeleRequest,
  TeleResponse,
  TeleResponseError,
  UnPromise,
} from './types'

const JSON_RPC_PARSE_ERROR = -32700
const JSON_RPC_INVALID_REQUEST = -32600
const JSON_RPC_METHOD_NOT_FOUND = -32601

// export function uniCallExpressMiddleware(context: unknown) {
//   return (req, res) => {
//     runWithContext(context, () => {
//       execute(req.body as TeleRequest<Fn>)
//         .then(({ status, json }) => res.status(status).json(json))
//         .catch((error) => res.status(ERROR).json(error))
//     })
//   }
// }

export default async function execute<T extends Fn>(
  request: TeleRequest<T>,
  api: Record<string, T>,
): Promise<TeleResponse<T>> {
  const [fn, errorValue] = getFn<T>(request, api)
  if (errorValue) return errorValue
  const { jsonrpc, params, id } = request
  try {
    // @ts-expect-error 2721 2488
    const result = (await fn(...params)) as UnPromise<ReturnType<T>>
    return { jsonrpc, id, result }
  } catch (error: unknown) {
    const {
      code = 0,
      message = 'Unknown error',
      data = JSON.stringify(error),
    } = error as TeleResponseError
    return { jsonrpc, id, error: { code, message, data } }
  }
}

function getFn<T extends Fn>(
  request: TeleRequest<T> | null,
  api: Record<string, T>,
): [null, TeleResponse<T>] | [T, null] {
  if (!request) return [null, parseError()]
  const { jsonrpc, method, id } = request
  const isInvalid = jsonrpc !== '2.0' || typeof method !== 'string' || id === undefined
  if (isInvalid) return [null, invalidRequest()]
  const fn = api[method] as T
  if (!fn) return [null, methodNotFound(request)]
  return [fn, null]
}

function parseError() {
  return {
    jsonrpc: '2.0' as const,
    error: { code: JSON_RPC_PARSE_ERROR, message: 'Parse error' },
    id: null,
  }
}

function invalidRequest() {
  return {
    jsonrpc: '2.0' as const,
    error: { code: JSON_RPC_INVALID_REQUEST, message: 'Invalid Request' },
    id: null,
  }
}

function methodNotFound<T extends Fn>({ jsonrpc, method, id }: TeleRequest<T>) {
  return {
    jsonrpc,
    error: {
      code: JSON_RPC_METHOD_NOT_FOUND,
      message: 'Method not found',
      data: { method },
    },
    id,
  }
}
