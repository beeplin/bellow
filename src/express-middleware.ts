import { Application, NextFunction, Request, Response } from 'express'
import path from 'path'
import { getConfig } from './utils'

export interface Context {
  req: Request
  res: Response
}

type Headers = Record<string, string>

interface Options {
  extraHeaders?: Headers
  convertResolverPath?: (input: string) => string
}

const { resolverBasePath, requestEndpoint } = getConfig()
const { port, pathname: requestBasePath } = requestEndpoint
const requestBasePathLength = requestBasePath.length

export default function (app: Application, options: Options = {}) {
  app.set('port', port)
  const { extraHeaders = {}, convertResolverPath = (s) => s } = options
  const headers = { 'Content-Type': 'application/json', ...extraHeaders }

  return async function (req: Request, res: Response, next: NextFunction) {
    if (req.method !== 'POST' || !req.path.startsWith(requestBasePath)) {
      next()
      return
    }
    if (!req.body) {
      next(new Error(`'app.use(express.json())' is missing.`))
      return
    }

    const relativePath = req.path.substr(requestBasePathLength)
    const location = path.resolve(resolverBasePath, convertResolverPath(relativePath))
    const func = await import(location).then((m) => m.default)
    const context: Context = { req, res }
    const result = await func(context, ...req.body)
    res.set(headers).status(200).json(result).end()

    return
  }
}
