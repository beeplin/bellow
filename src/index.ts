export * from './context'

export * from './error'

export * from './execute'

export * from './express'

export * from './session'

export * from './types'
