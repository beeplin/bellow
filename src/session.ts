/* eslint-disable max-classes-per-file */
import type express from 'express'

export interface IdGenerator {
  generate: () => string
}

export class FakeIdGenerator implements IdGenerator {
  private static id = 1000

  generate(): string {
    FakeIdGenerator.id += 1
    return String(FakeIdGenerator.id)
  }
}

export abstract class Session {
  protected _id: string | null = null

  get id(): string | null {
    return this._id
  }

  abstract changeId(): void
}

export class FakeSession extends Session {
  constructor(private readonly idGenerator: IdGenerator = new FakeIdGenerator()) {
    super()
    this.changeId()
  }

  changeId(): void {
    this._id = this.idGenerator.generate()
  }
}

export class ExpressSession extends Session {
  constructor(
    private readonly req: express.Request,
    private readonly res: express.Response,
    private readonly idGenerator: IdGenerator = new FakeIdGenerator(),
  ) {
    super()
    const id = this.req.headers.authorization
    if (id) this._id = id
    else this.changeId()
  }

  changeId(): void {
    this._id = this.idGenerator.generate()
    this.res.setHeader('authorization', this._id)
  }
}
