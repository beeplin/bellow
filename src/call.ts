/// <reference lib="dom" />

import 'isomorphic-fetch'
import { TeleError } from './error'
import type {
  Fn,
  PromiseReturnType,
  TeleInfo,
  TeleRequest,
  TeleResponse,
} from './types'

const JSON_RPC_INVALID_REQUEST = -32600

let callId = 0

export default call

async function call<T extends Fn>(
  fn: T | TeleInfo,
  ...params: Parameters<T>
): PromiseReturnType<T> {
  if (typeof fn === 'function') return fn(...params)
  nextId()
  const { endpoint, method, sessionTokenPersistence } = fn
  const request: TeleRequest<T> = {
    jsonrpc: '2.0',
    method,
    params,
    id: callId,
  }
  const rawResponse = await fetch(endpoint, {
    method: 'POST',
    headers: buildHeadersByPersistedToken(endpoint, sessionTokenPersistence),
    body: JSON.stringify(request),
    credentials: 'include',
  })
  persistTokenFromResponseHeaders(
    rawResponse.headers,
    endpoint,
    sessionTokenPersistence,
  )
  const response = (await rawResponse.json()) as TeleResponse<T>
  handleErrors(response, request)
  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
  return response.result!
}

function nextId() {
  callId = callId >= Number.MAX_SAFE_INTEGER ? Number.MIN_SAFE_INTEGER : callId + 1
}

const memoryStorage: Record<string, string> = {}

function buildHeadersByPersistedToken(
  endpoint: string,
  sessionTokenPersistence?: string,
) {
  const headers = new Headers({ 'content-type': 'application/json' })
  if (!sessionTokenPersistence || sessionTokenPersistence === 'localStorage') {
    const tokenName = getTokenName(endpoint)
    const token =
      globalThis.localStorage?.getItem(tokenName) ?? memoryStorage[tokenName] ?? '' // TODO cookie fetch
    headers.set('authorization', token)
  }
  return headers
}

function persistTokenFromResponseHeaders(
  headers: Headers,
  endpoint: string,
  sessionTokenPersistence?: string,
) {
  if (!sessionTokenPersistence || sessionTokenPersistence === 'localStorage') {
    const token = headers.get('authorization')
    if (token != null) {
      const tokenName = getTokenName(endpoint)
      if (globalThis.localStorage) localStorage.setItem(tokenName, token)
      else memoryStorage[tokenName] = token
    }
  }
}

function handleErrors<T extends Fn>(
  response: TeleResponse<T>,
  request: TeleRequest<T>,
) {
  if (response.error) throw new TeleError(response.error)
  if (response.jsonrpc !== '2.0' || response.id !== request.id)
    throw new TeleError({
      code: JSON_RPC_INVALID_REQUEST,
      message: 'Invalid Request',
    })
}

function getTokenName(endpoint: string) {
  return `sessionToken::${endpoint}`
}
