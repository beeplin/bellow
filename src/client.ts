/// <reference lib="dom" />

import { TeleError } from './error'
import type {
  Client,
  Fn,
  PromiseReturnType,
  RestParams,
  TeleRequest,
  TeleResponse,
} from './types'

const JSON_RPC_INVALID_REQUEST = -32600

export class TeleClient implements Client {
  fetch: typeof window.fetch

  headers: Headers = new Headers({ 'content-type': 'application/json' })

  id = 0

  // eslint-disable-next-line max-params
  constructor(
    fetch: typeof window.fetch,
    readonly endpoint: string,
    readonly tokenPersistence: typeof window.localStorage | 'cookie' = 'cookie',
    readonly localStorageName: string = 'sessionToken',
    readonly getMethodName = (fn: { name: string; path: string }) =>
      `${fn.path}//${fn.name}`,
  ) {
    this.fetch = fetch === globalThis.fetch ? fetch.bind(globalThis) : fetch
  }

  async call<T extends Fn>(fn: T, ...params: RestParams<T>): PromiseReturnType<T> {
    this.nextId()
    this.addTokenToRequest()
    const request: TeleRequest<T> = {
      jsonrpc: '2.0',
      // @ts-expect-error 2345
      method: this.getMethodName(fn),
      params,
      id: this.id,
    }
    const response = await this.fetch(this.endpoint, {
      method: 'POST',
      headers: this.headers,
      body: JSON.stringify(request),
    })
    this.persistTokenFromResponse(response)
    const { jsonrpc, result, error, id } = (await response.json()) as TeleResponse<T>
    if (error) throw new TeleError(error)
    if (jsonrpc !== '2.0' || id !== request.id || result === undefined)
      throw new TeleError({
        code: JSON_RPC_INVALID_REQUEST,
        message: 'Invalid Request',
      })
    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
    return result
  }

  reset(): void {
    if (this.tokenPersistence === 'cookie') return
    this.tokenPersistence.removeItem(this.localStorageName)
  }

  private addTokenToRequest() {
    if (this.tokenPersistence === 'cookie') return
    const token = this.tokenPersistence.getItem(this.localStorageName) ?? ''
    this.headers.set('authorization', token)
  }

  private persistTokenFromResponse(response: Response) {
    if (this.tokenPersistence === 'cookie') return
    const token = response.headers.get('authorization')
    if (token != null) this.tokenPersistence.setItem(this.localStorageName, token)
  }

  private nextId() {
    this.id = this.id >= Number.MAX_SAFE_INTEGER ? Number.MIN_SAFE_INTEGER : this.id + 1
  }
}
