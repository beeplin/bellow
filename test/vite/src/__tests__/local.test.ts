import { FakeSession } from '../../../../src/session'
import context0 from '../../../server0/src/context'
import context1 from '../../../server1/src/context'
import { run } from '../model'

describe('local run', () => {
  it('works', async () => {
    context0.set({ server: 'local0', session: new FakeSession() })
    context1.set({ server: 'local1', session: new FakeSession() })
    const res = await run()

    expect(res).toMatchSnapshot()
  })
})
