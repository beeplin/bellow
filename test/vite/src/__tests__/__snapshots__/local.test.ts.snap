// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`local run works 1`] = `
Array [
  Object {
    "id": null,
    "method": "echo",
    "name": "0",
    "server": "local0",
  },
  Object {
    "id": "AAA",
    "method": "login",
    "server": "local0",
  },
  Object {
    "id": "AAA",
    "method": "echo",
    "name": "2",
    "server": "local0",
  },
  Object {
    "id": "AAA",
    "method": "echo1",
    "res": Object {
      "id": null,
      "method": "echo",
      "name": "a",
      "server": "local1",
    },
    "server": "local0",
  },
  Object {
    "id": "AAA",
    "method": "login1",
    "res": Object {
      "id": "BBB",
      "method": "login",
      "server": "local1",
    },
    "server": "local0",
  },
  Object {
    "id": "AAA",
    "method": "echo1",
    "res": Object {
      "id": "BBB",
      "method": "echo",
      "name": "b",
      "server": "local1",
    },
    "server": "local0",
  },
  Object {
    "id": "",
    "method": "logout",
    "server": "local0",
  },
  Object {
    "id": "",
    "method": "echo",
    "name": "4",
    "server": "local0",
  },
  Object {
    "id": "",
    "method": "echo1",
    "res": Object {
      "id": "BBB",
      "method": "echo",
      "name": "c",
      "server": "local1",
    },
    "server": "local0",
  },
  Object {
    "id": "",
    "method": "logout1",
    "res": Object {
      "id": "",
      "method": "logout",
      "server": "local1",
    },
    "server": "local0",
  },
  Object {
    "id": "",
    "method": "echo1",
    "res": Object {
      "id": "",
      "method": "echo",
      "name": "d",
      "server": "local1",
    },
    "server": "local0",
  },
  Object {
    "id": "CCC",
    "method": "login",
    "server": "local0",
  },
  Object {
    "id": "CCC",
    "method": "login1",
    "res": Object {
      "id": "EEE",
      "method": "login",
      "server": "local1",
    },
    "server": "local0",
  },
  Array [
    Object {
      "id": "DDD",
      "method": "login",
      "server": "local1",
    },
    Object {
      "id": "DDD",
      "method": "echo",
      "name": "echo",
      "server": "local1",
    },
  ],
]
`;
